parsec-clients/src/main/java/com/yahoo/parsec/clients/ParsecResponseFuture.java

```diff
@@ -0,0 +1,155 @@
+        this.expectedStatusCodes = expectedStatusCodes;
+    }
+
+    /**
+     * Constructor.
+     * @param future future.
+     * @param tClass class type.
+     * @param expectedStatusCodes expected status codes.
+     */
+    public ParsecResponseFuture(Future<Response> future, Class<T> tClass,
+                                Set<Integer> expectedStatusCodes) {
+        this(future, tClass, expectedStatusCodes, new ObjectMapper());
+    }
+
+    /**
+     * Get the response.
+     * @return Response response
+     * @throws ExecutionException Execution exception
+     * @throws InterruptedException Interrupted exception
+     */
+    public Response getResponse() throws ExecutionException, InterruptedException {
+        return future.get();
+    }
+
+    /**
+     * Cancel.
+     *
+     * @param mayInterruptIfRunning mayInterruptIfRunning
+     * @return cancel was success or not
+     */
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        return future.cancel(mayInterruptIfRunning);
+    }
+
+    /**
+     * Is cancelled.
+     *
+     * @return boolean
+     */
+    @Override
+    public boolean isCancelled() {
+        return future.isCancelled();
+    }
+
+    /**
+     * Is done.
+     *
+     * @return boolean
+     */
+    @Override
+    public boolean isDone() {
+        return future.isDone();
+    }
+
+    /**
+     * Get the response object.
+     * @return T
+     * @throws ExecutionException Execution exception.
+     * @throws InterruptedException Interrupted exception.
+     * @throws HTTPException Http exception.
+     */
+    @Override
+    public T get() throws ExecutionException, InterruptedException, HTTPException {
+        Response response = getResponse();
+        int statusCode = response.getStatus();
+
+        if (expectedStatusCodes != null && expectedStatusCodes.contains(statusCode)) {
+            try {
+                if (response.hasEntity()) {
+                    String body = response.getEntity().toString();
+                    return objectMapper.readValue(body, tClass);
+                }
+                return null;
+            } catch (IOException e) {
+                throw new RuntimeException("Invalid Response Body");
+            }
+        } else {
+            throw new HTTPException(statusCode);
+        }
+    }
+
+    /**
+     * Get the response object with timeout.
+     * @param timeout timeout.
+     * @param unit time unit.
+     * @return T
+     * @throws ExecutionException Execution exception.
+     * @throws InterruptedException Interrupted exception.
+     * @throws TimeoutException Timeout exception.
+     * @throws HTTPException Http exception.
+     */
+    @Override
+    public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException,
+            TimeoutException, HTTPException {
+        // make sure that we can get the response before timeout
+        future.get(timeout, unit);
+        return get();
```

Comment:
maybe it's just me, I found this implementation is not as straight forward. I'd suggest to refactor get(long, TimeUnit) and get() a bit